group-coordinator/src/main/java/org/apache/kafka/coordinator/group/consumer/ConsumerGroup.java

```diff
@@ -516,16 +522,78 @@ public Assignment targetAssignment(String memberId) {
         return targetAssignment.getOrDefault(memberId, Assignment.EMPTY);
     }
 
+    /**
+     * @return An immutable map containing all the topic partitions
+     *         with their current member assignments.
+     */
+    public Map<Uuid, Map<Integer, String>> partitionAssignments() {
+        return Collections.unmodifiableMap(partitionAssignments);
+    }
+
     /**
      * Updates target assignment of a member.
      *
      * @param memberId              The member id.
      * @param newTargetAssignment   The new target assignment.
      */
     public void updateTargetAssignment(String memberId, Assignment newTargetAssignment) {
+        updatePartitionAssignments(
+            memberId,
+            targetAssignment.getOrDefault(memberId, new Assignment(Collections.emptyMap())),
+            newTargetAssignment
+        );
         targetAssignment.put(memberId, newTargetAssignment);
     }
 
+    /**
+     * Updates partition assignments of the topics.
+     *
+     * @param memberId              The member Id.
+     * @param oldTargetAssignment   The old target assignment.
+     * @param newTargetAssignment   The new target assignment.
+     *
+     * Package private for testing.
+     */
+    void updatePartitionAssignments(
+        String memberId,
+        Assignment oldTargetAssignment,
+        Assignment newTargetAssignment
+    ) {
+        // Combine keys from both old and new assignments.
+        Set<Uuid> allTopicIds = new HashSet<>();
+        allTopicIds.addAll(oldTargetAssignment.partitions().keySet());
+        allTopicIds.addAll(newTargetAssignment.partitions().keySet());
+
+        for (Uuid topicId : allTopicIds) {
+            Set<Integer> oldPartitions = oldTargetAssignment.partitions().getOrDefault(topicId, Collections.emptySet());
+            Set<Integer> newPartitions = newTargetAssignment.partitions().getOrDefault(topicId, Collections.emptySet());
+
+            TimelineHashMap<Integer, String> topicPartitionAssignment = partitionAssignments.computeIfAbsent(
+                topicId, k -> new TimelineHashMap<>(snapshotRegistry, Math.max(oldPartitions.size(), newPartitions.size()))
+            );
+
+            // Remove partitions that aren't present in the new assignment.
+            for (Integer partition : oldPartitions) {
+                if (!newPartitions.contains(partition) && memberId.equals(topicPartitionAssignment.get(partition))) {
```

Comment:
on `memberId.equals(topicPartitionAssignment.get(partition))`: will this ever be false? it seems we would always assign a partition to a member only when it's been removed by the previous member, right?

Reply:
Imagine there are two members A,B. We remove a partition (P0) from A and assign it to B. If B heartbeats first and we see that it's new target assignment has P0 now we will update the map to 1,B. This way when A heartbeats and we don't remove it from the map unless it's currently assigned to A still.
In the case where we just use a byte array it was possible that we update the assignment as 1,true and then unset it when A heartbeats even though it is currently assigned to B now

Reply:
It may be worth putting a comment about this.

Reply:
to confirm, `1,B` is referring to `P0 -> B` right? and what is setting and unsetting the assignment to true/false?

i think i understand. we need to either ensure that 1) we remove the partition if it's the same member id or 2) we add a partition if there are no member ids assigned to the partition

Reply:
yes my bad I meant P0,B

Reply:
That was our original implementation strategy, explained it so that it makes sense why we use a partition to memberId map now and check for the memberId before we remove a partition